File: csharp/Resources/GroupsResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Authdog.Types;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Directory group Wave 1 operations
    /// </summary>
    public class GroupsResource
    {
        private readonly AuthdogClient _client;

        public GroupsResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> CreateAsync(object body) =>
            _client.RequestAsync<JObject>(HttpMethod.Post, "/v1/groups", body);

        public JObject Create(object body) => CreateAsync(body).GetAwaiter().GetResult();

        public Task<EnvGroupsResponse> ListAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<EnvGroupsResponse>(
                HttpMethod.Get,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/groups");

        public EnvGroupsResponse List(string tenantId, string environmentId) =>
            ListAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> DeleteAsync(string tenantId, string environmentId, string groupId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/groups/{groupId}");

        public JObject Delete(string tenantId, string environmentId, string groupId) =>
            DeleteAsync(tenantId, environmentId, groupId).GetAwaiter().GetResult();

        public Task<JObject> ListMembersAsync(string tenantId, string environmentId, string groupId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/groups/{groupId}/members");

        public JObject ListMembers(string tenantId, string environmentId, string groupId) =>
            ListMembersAsync(tenantId, environmentId, groupId).GetAwaiter().GetResult();

        public Task<JObject> AddMemberAsync(string tenantId, string environmentId, string groupId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/groups/{groupId}/members",
                body);

        public JObject AddMember(string tenantId, string environmentId, string groupId, object body) =>
            AddMemberAsync(tenantId, environmentId, groupId, body).GetAwaiter().GetResult();

        public Task<JObject> RemoveMemberAsync(
            string tenantId,
            string environmentId,
            string groupId,
            string userId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"/v1/tenants/{tenantId}/environments/{environmentId}/groups/{groupId}/members/{userId}");

        public JObject RemoveMember(string tenantId, string environmentId, string groupId, string userId) =>
            RemoveMemberAsync(tenantId, environmentId, groupId, userId).GetAwaiter().GetResult();
    }
}

File: java/src/main/java/com/authdog/resources/GroupsResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.authdog.types.EnvGroupsResponse;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Directory group operations.
 */
public final class GroupsResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create a groups helper.
     * @param clientParam Authdog client
     */
    public GroupsResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * Create a group.
     * @param body request body
     * @return created group envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode create(final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST", "/v1/groups", body, null,
                JsonNode.class);
    }

    /**
     * List groups in an environment.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return groups envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public EnvGroupsResponse list(final String tenantId,
                                  final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/groups",
                null, null, EnvGroupsResponse.class);
    }

    /**
     * Delete a group.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param groupId group ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode delete(final String tenantId,
                           final String environmentId,
                           final String groupId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/groups/" + groupId,
                null, null, JsonNode.class);
    }

    /**
     * List members of a group.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param groupId group ID
     * @return members envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listMembers(final String tenantId,
                                final String environmentId,
                                final String groupId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/groups/" + groupId
                        + "/members",
                null, null, JsonNode.class);
    }

    /**
     * Add a user to a group.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param groupId group ID
     * @param body request body
     * @return add-member envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode addMember(final String tenantId,
                              final String environmentId,
                              final String groupId, final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/groups/" + groupId
                        + "/members",
                body, null, JsonNode.class);
    }

    /**
     * Remove a user from a group.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param groupId group ID
     * @param userId user ID
     * @return remove envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode removeMember(final String tenantId,
                                 final String environmentId,
                                 final String groupId,
                                 final String userId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                "/v1/tenants/" + tenantId + "/environments/"
                        + environmentId + "/groups/" + groupId
                        + "/members/" + userId,
                null, null, JsonNode.class);
    }
}
